Samples: Bluetooth: Mesh: light nRF52 DFU support

Andrewpini opened the pull request:
Adds point to point DFU support for nRF52 series, over SMP, to the mesh light sample.

Signed-off-by: Anders Storrø <anders.storro@nordicsemi.no>

Andrewpini commented:
Will start adding docs soon

NordicBuilder commented:
### Integration test specification

| Test Module | File based changes | Manually selected | West overwrite |
| ----------- | :----------------: | :---------------: | :------------: |
| test-fw-nrfconnect-ble_mesh | X | | |


[Detailed information of selected test modules](https://jenkins-ncs.nordicsemi.no/job/latest/job/sub/job/test-fw-nrfconnect-nrf/job/master/28021/artifact/test-spec/test-spec-summary.yml/*view*/)

*Note: This message is automatically posted and updated by the CI*

Balaklaka commented:
Remember to add something to release notes.

PavelVPV commented on samples/bluetooth/mesh/light/src/smp_dfu.c:
```diff
@@ -0,0 +1,62 @@
+
+static void disconnected(struct bt_conn *conn, uint8_t reason)
+{
+	k_work_submit(&advertise_work);
```
A false schedule may happen if disconnect happens for Mesh Provisioning or Mesh Proxy:
```
Bluetooth initialized
Mesh initialized
SMP BT service advertising successfully started
I: Primary Element: 0x0002
W: opcode 0x2039 status 0x09
SMP BT service advertising failed to start (rc -120)
```
The simplest fix I think would be to check error code of `bt_le_adv_start` against `-EALREADY`. But maybe the more correct way would be to use extended advertising API here and handle this somehow using `struct bt_le_ext_adv_cb` callbacks.

Andrewpini replied:
So I guess that is an issue in our other sample using this as well then?

PavelVPV replied:
Yes

Andrewpini replied:
> A false schedule may happen if disconnect happens for Mesh Provisioning or Mesh Proxy:
> 
> ```
> Bluetooth initialized
> Mesh initialized
> SMP BT service advertising successfully started
> I: Primary Element: 0x0002
> W: opcode 0x2039 status 0x09
> SMP BT service advertising failed to start (rc -120)
> ```
> 
> The simplest fix I think would be to check error code of `bt_le_adv_start` against `-EALREADY`. But maybe the more correct way would be to use extended advertising API here and handle this somehow using `struct bt_le_ext_adv_cb` callbacks.

I've tried and successfully added the extended advertiser to smp_dfu.c. Not sure how this can be used to avoid redundant advertisment of smp though. As far as I can see the `struct bt_le_ext_adv_cb` offers a `sent`, `connected` and `scanned` cb. At least the `connected` CB triggers for both device manager app and for the prov/proxy connections in the mesh app.

Andrewpini replied:
Maybe the 'sent` callback can be used 🤔

Andrewpini replied:
I have made an attempt using the `sent` CB. It seems to work Ok in testing. Review it and see what you think about it

PavelVPV commented on samples/bluetooth/mesh/light/src/main.c:
```diff
@@ -51,4 +52,8 @@ void main(void)
 	}
+
+	if (IS_ENABLED(CONFIG_BT_MESH_LIGHT_SAMPLE_NRF52_DFU)) {
+		smp_dfu_init();
```
Could `SYS_INIT` be used here like in `thingy53.c`?

Andrewpini replied:
It could, but does not work out of the box. Trying to run it with `k_work` makes the advertising fail. Using `k_work_delayable` however with a seconds delay seems to work. Not sure if this is a little to ad hoc solution. What do you think @PavelVPV ?

PavelVPV commented on samples/bluetooth/mesh/light/Kconfig:
```diff
@@ -0,0 +1,21 @@
+
+config BT_MESH_LIGHT_SAMPLE_NRF52_DFU
+	bool "Enable SMP DFU support"
+	depends on MCUMGR_CMD_OS_MGMT
+	depends on MCUMGR_CMD_IMG_MGMT
+	depends on MCUMGR_SMP_BT
```
Aren't these options enabled in overlay file?

Andrewpini replied:
Yes, they are. But `depends on` does not pick the option as far as I know, it only makes the compiler deselect the dependent option unless all `depends on` requirements are fulfilled. I choose those dependencies since they are required to build `smp_dfu.c`

PavelVPV replied:
But aren't they enabled if use overlay file? How do you see a user should enable dfu: through overlay file or through menuconfig, or somehow else?

Andrewpini replied:
I guess it is just redundancy. I want the user to use the overlay to enable dfu

Andrewpini replied:
I can remove it

PavelVPV commented on samples/bluetooth/mesh/light/Kconfig:
```diff
@@ -0,0 +1,21 @@
+	depends on MCUMGR_CMD_OS_MGMT
+	depends on MCUMGR_CMD_IMG_MGMT
+	depends on MCUMGR_SMP_BT
+	depends on SOC_SERIES_NRF52X
```
I've no strong opinion on this, but what if we don't enforce 52 series here. Then it should be possible to run this code with nRF53DK. I guess `smp_dfu.c` could be conditionally compiled in if, for example, `CONFIG_MCUMGR_SMP_BT` is enabled.

Andrewpini replied:
I get your point. The issue is Thingy 53. I wanted to create a clear distinction between the dfu for thingy53 and the other platforms. Picking 52 series was the most easy option (only excluding `nrf5340dk_nrf5340`)

Andrewpini replied:
I answered Ingar on this bellow ⬇️

Balaklaka commented on samples/bluetooth/mesh/light/Kconfig:
```diff
@@ -0,0 +1,21 @@
+
+menu "Bluetooth mesh light sample"
+
+config BT_MESH_LIGHT_SAMPLE_NRF52_DFU
```
Do we really need this config option. Can we solve it some other way by using some already existing configuration option that is in the overlay-dfu.conf file.

Andrewpini replied:
Well, if we want all platforms to run this, except thingy53, we would need to exclude `CONFIG_BOARD_THINGY53_NRF5340_CPUAPP
` and `CONFIG_BOARD_THINGY53_NRF5340_CPUAPP_NS` when running the init of smp_dfu. That would work, but either way we turn this it will look a bit ugly if we are not altering the board files for thingy53

Balaklaka commented on samples/bluetooth/mesh/light/src/main.c:
```diff
@@ -51,4 +52,8 @@ void main(void)
 		printk("Bluetooth init failed (err %d)\n", err);
 	}
+
+	if (IS_ENABLED(CONFIG_BT_MESH_LIGHT_SAMPLE_NRF52_DFU)) {
```
Could this use CONFIG_MCUMGR_SMP_BT instead? Or will that add thingy:53 too. Or could it then && !thingy:53

Andrewpini replied:
Seee above comment ⬆️

Andrewpini commented on samples/bluetooth/mesh/light/CMakeLists.txt:
```diff
@@ -18,6 +18,10 @@ target_sources(app PRIVATE
 target_include_directories(app PRIVATE include)
 
+# Preinitialization related to DFU over SMP for nRF52 series
+target_sources_ifdef(CONFIG_BT_MESH_LIGHT_SAMPLE_NRF52_DFU app PRIVATE
```
It requires additional configuration. Can we skip that platform for now just not to complicate this task any more? I guess the support for 52 series is what the customers mainly are interested in

PavelVPV replied:
Excluding compilation of `smp_dfu.c` by `CONFIG_BOARD_THINGY53_NRF5340_CPUAPP`/`_NS` maybe a good option. I was also thinking about taking another sample, which doesn't support thingy53, but they are not much in demand.

Andrewpini replied:
Using another sample would solve all problems, except that it would be hard for users to discover it unless we slap the feature in their face through documentation page. Using one of the light related samples is most likely to make users aware of this feature.

Andrewpini commented on samples/bluetooth/mesh/light/CMakeLists.txt:
```diff
@@ -18,6 +18,12 @@ target_sources(app PRIVATE
 target_include_directories(app PRIVATE include)
 
+# Preinitialization related to DFU over SMP for nRF52 series
+if(CONFIG_SOC_SERIES_NRF52X)
```
It also requires additional configurtaion as far as I have seen

Andrewpini replied:
If we want support for 53DK it requires a separate configuration for the network core that allows using the extended advertiser. Since legacy adv at the moment is the default, such a network core configuration should be conditionally added when using the overlay-dfu.conf. I talked to @m-alperen-sener about it, and while it is possible im concerned that it will be a bit convoluted if we want to add support for 53DK now. When we start using extended advertiser by default I think that this won't be a problem anymore.

PavelVPV replied:
Ok, lets leave it as is now.

Andrewpini commented on samples/bluetooth/mesh/light/README.rst:
```diff
@@ -145,6 +182,38 @@ Configure the Generic OnOff Server model on each element on the **Mesh Light** n
+* Flash the sample to the board.
+* **(Optional)** Provision and configure the mesh device.
+* To observe the application log, open a terminal program and connect to the device.
+* **(Optional)** Do some alterations to the sample and re-build it, i.e. add a print statement to the main function.
```
I did not have to to make it work at least, so it is not a hard requirement for the current solution at least. But maybe it should be?

PavelVPV replied:
Hmm, lets leave this as is.

Andrewpini commented on samples/bluetooth/mesh/light/README.rst:
```diff
@@ -145,6 +182,38 @@ Configure the Generic OnOff Server model on each element on the **Mesh Light** n
 
+Running DFU
+===========
+To perform a DFU, execute the following steps:
```
Makes sence, but I'll leave all documentation decisions up to @mia-ko.

PavelVPV replied:
Ok, but I'm not a fun of re-creating such instructions in everywhere. Having them in one place makes easier to support in future if the mobile application logic, or update archive file name changes. If we feel luck of steps in the main guide, then it should be updated instead.

Andrewpini commented on doc/nrf/links.txt:
```diff
@@ -849,6 +849,9 @@
 .. _`nRF Mesh mobile app for iOS`: https://apps.apple.com/us/app/nrf-mesh/id1380726771
 .. _`nRF Mesh mobile app for Android`: https://play.google.com/store/apps/details?id=no.nordicsemi.android.nrfmeshprovisioner
 
+.. _`nRF Device Manager mobile app for iOS`: https://apps.apple.com/us/app/nrf-connect-device-manager/id1519423539
```
See doc comment bellow ⬇️

Andrewpini commented on samples/bluetooth/mesh/light/README.rst:
```diff
@@ -30,6 +32,27 @@ The sample also requires a smartphone with Nordic Semiconductor's nRF Mesh mobil
+While this overlay configuration is only applicable for the nRF52 series in this sample, DFU over SMP can be utilized on other platforms as well.
+
+.. note::
+   Point-to-point DFU for :ref:`zephyr:thingy53_nrf5340` is supported by default.
```
See doc comment bellow ⬇️

PavelVPV replied:
Ok, but IMHO it would be user friendly to add the link so that users don't need to search for this in the documentation.

mia-ko commented on samples/bluetooth/mesh/light/README.rst:
```diff
@@ -145,6 +182,38 @@ Configure the Generic OnOff Server model on each element on the **Mesh Light** n
 
+Running DFU
+===========
+To perform a DFU, execute the following steps:
```
```suggestion
After the sample is built with the :file:`overlay-dfu.conf` and programmed to your development kit, support for FOTA upgrades is enabled.
See FOTA in Bluetooth mesh (link) for instructions on how to perform FOTA upgrade and initiate the DFU process.
```
I guess this would replace everything else you wrote in this section.

I would recommend saving your detailed steps in case we need them later, i.e. if we want to add some details either in the existing FOTA over BLE guide or in our own mesh guide regarding FOTA upgrades.

mia-ko commented on doc/nrf/releases/release-notes-changelog.rst:
```diff
@@ -206,6 +206,12 @@ Bluetooth mesh samples
+
+  * Added:
+
+    * Point-to-point DFU support for nRF52 boards over SMP.
```
```suggestion
    * Point-to-point Device Firmware Update (DFU) support over the Simple Management Protocol (SMP) for supported nRF52 Series development kits.
```

mia-ko commented on samples/bluetooth/mesh/light/README.rst:
```diff
@@ -30,6 +32,27 @@ The sample also requires a smartphone with Nordic Semiconductor's nRF Mesh mobil
+While this overlay configuration is only applicable for the nRF52 series in this sample, DFU over SMP can be utilized on other platforms as well.
+
+.. note::
+   Point-to-point DFU for :ref:`zephyr:thingy53_nrf5340` is supported by default.
```
```suggestion
   Point-to-point DFU for :ref:`zephyr:thingy53_nrf5340` is supported by default.
   See :ref:`thingy53_app_update` for more information about updating firmware image on :ref:`zephyr:thingy53_nrf5340`.
```
I wonder if it is better to link to https://developer.nordicsemi.com/nRF_Connect_SDK_dev/doc/PR-9109/nrf/ug_thingy53.html#updating-firmware-image than to https://developer.nordicsemi.com/nRF_Connect_SDK_dev/doc/PR-9109/nrf/ug_thingy53_gs.html#getting-started-with-precompiled-firmware-samples. What do you think?

Andrewpini commented:
I've added the suggestions from @mia-ko. Need to add the link to the detailed FOTA guide when it is ready. I also moved the init of smp_dfu to the `bt_ready` CB. The reason is that I've tested the implementation for nrf53DK just to see how it should be done, and it was required to move to make it initialize properly on that platform. As soon as we have cleaned up the SMP dfu stuff for all platforms it should be fairly easy to make the same dfu init apply to all platforms.

PavelVPV commented on samples/bluetooth/mesh/light/src/main.c:
```diff
@@ -51,4 +56,5 @@ void main(void)
 	if (err) {
 		printk("Bluetooth init failed (err %d)\n", err);
 	}
+
```
:)

mia-ko reviewed (approved):
Approving this, and once it is merged I will edit samples/bluetooth/mesh/light/README.rst to add the missing link to the FOTA in Bluetooth mesh in my PR.

ludvigsj commented on samples/bluetooth/mesh/light/README.rst:
```diff
@@ -145,6 +183,12 @@ Configure the Generic OnOff Server model on each element on the **Mesh Light** n
+===========
+
+After the sample is built with the :file:`overlay-dfu.conf` and programmed to your development kit, support for FOTA upgrades is enabled.
+See FOTA in Bluetooth mesh (link) for instructions on how to perform FOTA upgrade and initiate the DFU process.
```
I assume this `(link)` is gonna be made into an actual link before this is merged?

ludvigsj replied:
Ah of course Mia explained this while I was reviewing. Then I guess it's fine

Andrewpini replied:
Mia will fix it 👍
